webui/react/src/hooks/useCustomizeColumnsModal.tsx

```diff
@@ -0,0 +1,234 @@
+    // We always render the form regardless of mode to provide a reference to it.
+    return (
+      <div className={css.base}>
+        <Input placeholder="Search columns..." onChange={handleSearch} />
+        <div className={css.columns}>
+          <div className={css.column}>
+            <h2>Hidden</h2>
+            <List
+              className={css.listContainer}
+              height={200}
+              innerElementType="ul"
+              itemCount={filteredHiddenColumns.length}
+              itemSize={24}
+              width="100%">
+              {renderHiddenRow}
+            </List>
+            <Button type="link" onClick={() => makeVisible(filteredHiddenColumns)}>
+              Add All
+            </Button>
+          </div>
+          <div className={css.column}>
+            <div className={css.visibleTitleRow}>
+              <h2>Visible</h2>
+              {!isEqual(state.defaultVisibleColumns, visibleColumns) && (
+                <Button type="link" onClick={() => resetColumns(state)}>
+                  Reset
+                </Button>
+              )}
+            </div>
+            <List
+              className={css.listContainer}
+              height={200}
+              innerElementType="ul"
+              itemCount={filteredVisibleColumns.length}
+              itemSize={24}
+              width="100%">
+              {renderVisibleRow}
+            </List>
+            <Button type="link" onClick={() => makeHidden(filteredVisibleColumns)}>
+              Remove All
+            </Button>
+          </div>
+        </div>
+      </div>
+    );
+  }, [ handleSearch,
+    filteredHiddenColumns,
+    renderHiddenRow,
+    visibleColumns,
+    filteredVisibleColumns,
+    renderVisibleRow,
+    makeVisible,
+    resetColumns,
+    makeHidden ]);
+
+  const generateModalProps = useCallback((state: ModalState): Partial<ModalFuncProps> => {
+    const modalProps: Partial<ModalFuncProps> = {
+      bodyStyle: { padding: 0 },
+      className: css.base,
+      closable: true,
+      content: generateModalContent(state),
+      icon: null,
+      maskClosable: true,
+      okText: 'Save',
+      onCancel: handleCancel,
+      onOk: () => handleSave(state),
+      title: 'Customize Columns',
+    };
+
+    return modalProps;
+  }, [ generateModalContent, handleCancel, handleSave ]);
+
+  // Detect modal state change and update.
+  useEffect(() => {
+    if (!modalState.visible) return;
+    const modalProps = generateModalProps(modalState);
+    if (modalRef.current) {
+      modalRef.current.update(prev => ({ ...prev, ...modalProps }));
+    } else {
+      modalRef.current = Modal.confirm(modalProps);
+    }
+  }, [ generateModalProps, modalState ]);
+
+  // When the component using the hook unmounts, remove the modal automatically.
+  useEffect(() => {
+    return () => {
+      if (!modalRef.current) return;
+      modalRef.current.destroy();
+      modalRef.current = undefined;
+    };
+  }, []);
```

Comment:
**suggestion:** I think this will be good to have inside `useModal`. I'm not sure what kind of impact it has inside there but might be worth moving there and trying it out. I think in general should be a good behavior for all modals. Also within `useModal` could just call `modalClose` instead of lines `225 ~ 227`.